### PR TITLE
 ipq40xx: add MikroTik cAP ac support

### PR DESCRIPTION
**Pull request by alaraun**
This adds support for the MikroTik RouterBOARD RBcAPGi-5acD2nD
(cAP ac), a  indoor dual band, dual-radio 802.11ac wireless AP, two
10/100/1000 Mbps Ethernet ports.

See https://mikrotik.com/product/cap_ac for more info.

Specifications:
 - SoC: Qualcomm Atheros IPQ4018
 - RAM: 128 MB
 - Storage: 16 MB NOR
 - Wireless:
   · Built-in IPQ4018 (SoC) 802.11b/g/n 2x2:2, 2.5 dBi antennae
   · Built-in IPQ4018 (SoC) 802.11a/n/ac 2x2:2, 2.5 dBi antennae
 - Ethernet: Built-in IPQ4018 (SoC, QCA8075) , 2x 1000/100/10 port,
   PoE in and passive PoE out

Installation:
Boot the initramfs image via TFTP and then flash the sysupgrade
image using "sysupgrade -n"

### CODE REVIEW

**Comment by john-tho** [comment]
> Thanks for your contribution to OpenWrt!
> 
> To help keep the codebase consistent and readable,
> and to help people review your contribution,
> we ask you to follow the rules you find in the wiki at this link
> https://openwrt.org/submitting-patches
> 
> Please remove this message before posting the pull request.

You must follow the submission guidelines. See signoff.
https://openwrt.org/submitting-patches#submission_guidelines

`--amend` your commit, and force push it to update this PR.

Do not duplicate code in case switches when there is another case with exactly the same code. (done in network mac addresses, wifi caldata, upgrade). Add an alternate case instead:
```
	mikrotik,cap-ac|\
	mikrotik,hap-ac2)
```

Do you want the wifi LEDs to do anything?
Does the PoE output have GPIOs for detection and control?
https://help.mikrotik.com/docs/display/UM/cAP+ac#heading-LEDindicators mentions additional LEDs for PoE in, and PoE out.
https://help.mikrotik.com/docs/display/ROS/LEDs mentions all LEDs can be turned off

I would like to see a shared ipq40xx mikrotik dtsi, but that may get done in the SXTsq 5 ac PR.

**Comment by alaraun** [comment]
> > Thanks for your contribution to OpenWrt!
> > To help keep the codebase consistent and readable,
> > and to help people review your contribution,
> > we ask you to follow the rules you find in the wiki at this link
> > https://openwrt.org/submitting-patches
> > Please remove this message before posting the pull request.
> 
> You must follow the submission guidelines. See signoff.
> https://openwrt.org/submitting-patches#submission_guidelines
> 
> `--amend` your commit, and force push it to update this PR.
Done.
> 
> Do not duplicate code in case switches when there is another case with exactly the same code. (done in network mac addresses, wifi caldata, upgrade). Add an alternate case instead:
> 
> ```
> 	mikrotik,cap-ac|\
> 	mikrotik,hap-ac2)
> ```
Done.

> 
> Do you want the wifi LEDs to do anything?
Now they are.

> Does the PoE output have GPIOs for detection and control?
I have added poeout led but I don't know the control part.
> https://help.mikrotik.com/docs/display/UM/cAP+ac#heading-LEDindicators mentions additional LEDs for PoE in, and PoE out.
> https://help.mikrotik.com/docs/display/ROS/LEDs mentions all LEDs can be turned off
Not sure how to solve this.

> 
> I would like to see a shared ipq40xx mikrotik dtsi, but that may get done in the SXTsq 5 ac PR.
Skipped

**Comment by john-tho** [comment]
Thanks. Looking better.

If you want to have functioning passive PoE out, we can try to work it out, otherwise list that it is not supported in the commit message.
It may be very different to the older devices I have seen, or just work.
https://help.mikrotik.com/docs/display/ROS/PoE-Out

The ath79 & ramips Mikrotik devices with passive PoE out that I have seen:
- use a GPIO in input mode to detect when a (possibly Mikrotik passive PoE) compatible device is connected to the PoE out port
- use a GPIO in output mode to switch the PoE out on/off
- given all the LEDs can be turned off, I would guess a different GPIO controls the LED to the PoE out

Possible ways to work it out:
- OEM `dmesg` or supout (`/system sup-output name=supout.rif`), then `decode-supout.py`
- OEM DTB (you can get this while still running OpenWrt on the device: save the dtb_config mtd partition), then `dtc` to convert to DTS, or OEM GPL source `platform.c`, or `board-.*.c`.
- OEM jailbreak, upload `busybox`, and watch `/sys/kernel/debug/gpio`, or the SoC register memory locations, before and after you do each:
  - Force PoE on/off through `/interface ethernet poe`
  - Plug/unplug a Mikrotik passive PoE compatible device (_resistance on spare pairs in the range from 3kΩ to 26.5kΩ_)

**Comment by alaraun** [comment]
> Thanks. Looking better.
> 
> If you want to have functioning passive PoE out, we can try to work it out, otherwise list that it is not supported in the commit message.

Will mark it as unsupported for now. I'll try to figure it out later.

> It may be very different to the older devices I have seen, or just work.
> https://help.mikrotik.com/docs/display/ROS/PoE-Out
> 
> The ath79 & ramips Mikrotik devices with passive PoE out that I have seen:
> 
>     * use a GPIO in input mode to detect when a (possibly Mikrotik passive PoE) compatible device is connected to the PoE out port
> 
>     * use a GPIO in output mode to switch the PoE out on/off
> 
>     * given all the LEDs can be turned off, I would guess a different GPIO controls the LED to the PoE out

In Mikrotik dts GPIO 0 is marked as all-leds but it works for PoE-In led only.

> 
> 
> Possible ways to work it out:
> 
>     * OEM firmware `dmesg` or supout (`/system sup-output name=supout.rif`), then `decode-supout.py`
> 
>     * OEM firmware DTB (you can get this while still running OpenWrt on the device: save the dtb_config mtd partition), then `dtc` to convert to DTS, or OEM GPL source `platform.c`, or `board-.*.c`.
> 
>     * OEM firmware jailbreak, upload `busybox`, and watch `/sys/kernel/debug/gpio`, or the SoC register memory locations, before and after you do each:
>       
>       * Force PoE on/off through `/interface ethernet poe`
>       * Plug/unplug a Mikrotik passive PoE compatible device (_resistance on spare pairs in the range from 3kΩ to 26.5kΩ_)

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts
Comment (john-tho) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts:197]: One too many zeroes in the node name.
Picked up in hap ac2 here: https://github.com/openwrt/openwrt/pull/3037#discussion_r573224727

Comment (alaraun) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts]: Updated

**Comment by alaraun** [comment]
Shared ipq4018 Mikrotik dtsi added.

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds
Comment (adschm) [inline target/linux/ipq40xx/base-files/etc/board.d/01_leds:63]: please use DT triggers instead.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts:33]: Do you need the DT label?

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts
Comment (adschm) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts:3]: The idea of a shared DTSI is that it's shared. You don't share anything here.

Either create a DTSI that is actually shared between this device and the hap-ac2 (if that's possible), or put all code for your device into the DTS.

For a shared DTSI, the hard point would be to find a good name then, as it probably won't be shared between all ipq4018 Mikrotik devices, so just calling it "mikrotik" would probably not be a good idea (unless I'm wrong here).

**Review by adschm — CHANGES_REQUESTED** [review]

**Comment by rusitschka** [comment]
I've added this PR to my [Docker based build](https://github.com/rusitschka/openwrt-mikrotik-hap-ac2) and it works on my cAP ac. I'm missing Network/Switch however and everything VLAN related. I used hAP ac2 image before on my cAP ac and it showed Network/Switch - but VLANs did not work with that, either.

File: target/linux/ipq40xx/base-files/etc/board.d/01_leds
Comment (alaraun) [reply target/linux/ipq40xx/base-files/etc/board.d/01_leds]: Done

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts
Comment (alaraun) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts]: DT label removed

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts
Comment (alaraun) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-cap-ac.dts]: qcom-ipq4018-mikrotik.dtsi removed

**Comment by vibornoff** [comment]
@alaraun I've played a bit with `cAP AC` using @robimarko 's `hAP AC2` #3037 and dig some helpful stuff from the stock flash.
According to the next part of decompiled DTB
```
		poe-out-simple {
			compatible = "rb,poe-out-simple";
			status = "ok";
			eth_port = <0x01>;
			gpo_on = <0x4e 0x02 0x00>;
			gpi_status = <0x4e 0x04 0x00>;
		};
```
`GPIO 4` is used to detect plugged PoE device. `GPIO 2` is used to enable PoE passthrough.

Additinally from my observations:
* `GPIO 0` is not for just blue LED, it controls both blue and red LEDs;
* Setting `GPIO 59` low turns down at least eth1 phy.

For now I'm stuck at finding out how to control ethernet LEDs in order to enable "Dark Mode".

**Comment by vibornoff** [comment]
> For now I'm stuck at finding out how to control ethernet LEDs in order to enable "Dark Mode".

ETH LEDs are now blinkable from userspace: https://github.com/alaraun/openwrt/pull/1

**Comment by rusitschka** [comment]
> > For now I'm stuck at finding out how to control ethernet LEDs in order to enable "Dark Mode".
> 
> ETH LEDs are now blinkable from userspace: [alaraun#1](https://github.com/alaraun/openwrt/pull/1)

"Dark mode" is working for me with that PR - all LEDs can be turned off. Thanks! @vibornoff I also saw that you added the setting for a switch. However I still do not see it in UI. Do I have to factory reset in order to see it?

**Comment by vibornoff** [comment]
> However I still do not see it in UI. Do I have to factory reset in order to see it?
@rusitschka that's my `git rebase` issue :-) Fix is force-pushed just now.

**Comment by rusitschka** [comment]
> > However I still do not see it in UI. Do I have to factory reset in order to see it?
> 
> @rusitschka that's my `git rebase` issue :-) Fix is force-pushed just now.

Thanks! I just did another build but still don't see "Switch" in UI. Any idea?

**Comment by vibornoff** [comment]
> Thanks! I just did another build but still don't see "Switch" in UI. Any idea?

`make distclean` ?

**Comment by g6094199** [comment]
> > For now I'm stuck at finding out how to control ethernet LEDs in order to enable "Dark Mode".
> 
> ETH LEDs are now blinkable from userspace: [alaraun#1](https://github.com/alaraun/openwrt/pull/1)

Hi,

i see there there are still 2 LEDs  missing imho.  I cant use the "dark mode" button LED and the LED left of the power LED (a circle w. arrow outwards ) via Luci.

sash

**Comment by vibornoff** [comment]
> i see there there is still a LED missing imho. I cant see/use/setup the LED left of the power LED (a circle w. arrow outwards ) via Luci.

This LED is not directly controllable via GPIO.There is
* `GPIO 2` controlling PoE passthrough
* and `GPIO 0` acting as a key for both "power" and "poe out" LEDs.

So this LED is on when both `GPIO 0` and `GPIO 2` set to `1`.

**Comment by g6094199** [comment]
ah, ok. i see. to bad.
thx anyway!

**Comment by vibornoff** [comment]
Working on PoE passthrough... Stay tuned :-)

**Comment by rusitschka** [comment]
Reset to factory defaults gave me Switch menu in UI. However whenever I add a second VLAN I'm losing connectivity. 

Switch shows: CPU (eth0, 1000baseT full-duplex) | CPU (eth1, 1000baseT full-duplex) | LAN (no link)

I only have physical Eth1 connected (the port with PoE in). This seems to be eth1 in UI. eth0 in UI is CPU? And LAN physical Eth2 with PoE out?

**Comment by vibornoff** [comment]
> Switch shows: CPU (eth0, 1000baseT full-duplex) | CPU (eth1, 1000baseT full-duplex) | LAN (no link)
> 
> I only have physical Eth1 connected (the port with PoE in). This seems to be eth1 in UI. eth0 in UI is CPU? And LAN physical Eth2 with PoE out?

cAP AC has dual CPU <-> Switch link

```
  CPU       switch0
+------+   +--------+
| eth0 |---| port 0 |
| eth1 |   | port 1 |
+------+   | port 2 |                Enclousure
   |       | port 3 |     PHY          label
   |       | port 4 |---ethphy3---< E2/PoE out
   +-------| port 5 |---ethphy4---< E1/PoE in
           +--------+
```

**Comment by robimarko** [comment]
> > Switch shows: CPU (eth0, 1000baseT full-duplex) | CPU (eth1, 1000baseT full-duplex) | LAN (no link)
> > I only have physical Eth1 connected (the port with PoE in). This seems to be eth1 in UI. eth0 in UI is CPU? And LAN physical Eth2 with PoE out?
> 
> cAP AC has dual CPU <-> Switch link

Well, actually this is not possible with the IPQ4018 at all.
SoC is connected to the switch port 0 via an internal 1Gbit full-duplex interface.
Ports 1 to 5 are then connected via switch port 6 which has a PSGMII PHY to the QCA8072 PSGMII dual PHY.
eth0 and eth1 are just driver magic, nothing more than that as the SoC only has one GMAC.

**Comment by g6094199** [comment]
> Reset to factory defaults gave me Switch menu in UI. 

sounds like an browser cache issue....

**Comment by rusitschka** [comment]
Thanks for your anwsers. Does any of this explain why I'm losing connectivity when I add a second VLAN in switch? 

Is there something special in physical interfaces of bridge I have to take care of? 

What works is eth1 ethernet in bridge and VLAN 1 in switch: eth0 off, eth1 tagged, LAN untagged. But when I add second VLAN connection on VLAN 1 drops.

**Comment by robimarko** [comment]
VLANs as far as I know are kinda broken on the IPQ40xx as the EDMA driver forces VLANs 1 and 2.
But, I don't use VLANs on any IPQ4k gear so I would not know the current state.

**Comment by rusitschka** [comment]
I got it working! Thanks @robimarko for pointing me to the right direction. I found [this](https://forum.openwrt.org/t/ipq40xx-switch-config-strangeness/32542/66) and discovered that EDMA driver automagically tags eth0 to VLAN 1 and eth1 to VLAN 2. I managed to change eth1 to tag to VLAN 1 as well via /etc/sysctl.conf:
```
net.edma.default_group2_vlan_tag=1
net.edma.default_wan_tag=1
```
That did the trick since untagged packets are now put to VLAN 1.

**Comment by vibornoff** [comment]
> Well, actually this is not possible with the IPQ4018 at all.
> SoC is connected to the switch port 0 via an internal 1Gbit full-duplex interface.
> Ports 1 to 5 are then connected via switch port 6 which has a PSGMII PHY to the QCA8072 PSGMII dual PHY.
> eth0 and eth1 are just driver magic, nothing more than that as the SoC only has one GMAC.

@robimarko does it mean that there is no 2x1Gb CPU<->Switch link? And is there a way to completely disable that weird "driver magic" falling back to plain old `eth0.1` and `eth0.2` for LAN and WAN respectively?

**Comment by robimarko** [comment]
Yes, there is no 2x1 Gb link at all.
You can attempt to disable it like MR33 VLAN0 in the DTS, but I have no idea will it work if single phy is not used.

**Comment by vibornoff** [comment]
> Yes, there is no 2x1 Gb link at all.
> You can attempt to disable it like MR33 VLAN0 in the DTS, but I have no idea will it work if single phy is not used.

Great thanks to @robimarko to pointing the right direction. After digging a lot into EDMA driver sources this DTS tweak did the trick:
```diff
                edma@c080000 {
                        status = "okay";
+                       qcom,poll_required = <0>;
+                       qcom,num_gmac = <1>;
                };
        };

+&gmac0 {
+       vlan_tag = <0 0x3f>;
+};
```
Setting `num_gmac=1` prevents from creating virtual `eth1` device.
Setting `vlan_tag=0` for all ports completely disables weird EDMA vlan tagging.
Switch device is now completely configurable with LuCi and VLANs work as expected.
https://github.com/vibornoff/openwrt/tree/mikrotik-cap-ac-wip

However this tweak lead to another issue.
Since `eth0` device now handles flows for both VLANs it becomes useless as a source for LED trigger events.
And there is no `eth1` device anymore so there is no source to trigger E1 LED.

I tried with `switch0` trigger for `green:eth1` LED.
Setting `port_mask=0x20` and `mode=link tx rx` sense only link event for switch port 5 (labeled as eth1/PoE in) -- led lits on when cable is inserted and doesn't blink when traffic flows.

Any thoughts on how to make leds blinking on tx/rx?

**Comment by robimarko** [comment]
It should be able to trigger of the switch port traffic

**Comment by millenium7** [comment]
I've tried compiling the firmware and I get errors. Have followed the instructions to the letter so I have NFI why
Can you compile the latest release and upload it?

**Comment by kmarzic** [comment]
Hi,
I think I solved smp_affinity for ath10k_ahb (2.4 GHz and 5 GHz).

```
root@capac:~# cat /etc/hotplug.d/net/40-net-smp-affinity
#!/bin/sh

[ "$ACTION" = add ] || exit

get_device_irq() {
        local device="$1"

        local line=$(grep -m 1 "${device}" /proc/interrupts)
        echo ${line} | sed 's/:.*//'
}

set_interface_core() {
        local core_mask="$1"
        local interface="$2"
        local device="$3"

        [ -z "${device}" ] && device="$interface"

        local irq=$(get_device_irq "$device")

        echo "${core_mask}" > /proc/irq/${irq}/smp_affinity
}

case "$(board_name)" in
mikrotik,cap-ac)
        set_interface_core 2 "ath10k_ahb" "200"
        set_interface_core 4 "ath10k_ahb" "201"
        ;;
esac

root@capac:~#
```

Please give suggestions for improvements.

**Comment by pgaskin** [comment]
@robimarko, I'm going to try building master with @kmarzic's patch for the hAP ac<sup>2</sup>. If it works, would it be possible to get that in before 21.02 is released, or is it too late for a backwards-incompatible change like that?

**Comment by robimarko** [comment]
Its not really up to me, but to maintainers rather.

**Comment by pgaskin** [comment]
@robimarko, I had another idea for a backwards-compatible workaround for the ac<sup>2</sup>.

I edited `/etc/board.json` to look like (note the new eth1 and wan stuff at the beginning of ports and roles):

<details>

```json
	"switch": {
		"switch0": {
			"enable": true,
			"reset": true,
			"ports": [
				{
					"num": 0,
					"device": "eth1",
					"need_tag": false,
					"want_untag": true
				},
				{
					"num": 5,
					"role": "wan"
				},
				{
					"num": 0,
					"device": "eth0",
					"need_tag": false,
					"want_untag": true
				},
				{
					"num": 1,
					"role": "lan",
					"index": 4
				},
				{
					"num": 2,
					"role": "lan",
					"index": 3
				},
				{
					"num": 3,
					"role": "lan",
					"index": 2
				},
				{
					"num": 4,
					"role": "lan",
					"index": 1
				},
			],
			"roles": [
				{
					"role": "wan",
					"ports": "5 0",
					"device": "eth1"
				},
				{
					"role": "lan",
					"ports": "1 2 3 4 0",
					"device": "eth0"
				}
			]
		}
	},
```

This is equivalent to making the line in /etc/board.d/02_network into `"0u@eth1" "5:wan" "0u@eth0" "1:lan:4" "2:lan:3" "3:lan:2" "4:lan:1"`.

Since I was manually editing my config so I didn't need to test it from scratch, I also made the switch config in /etc/config/network into:

<details>

```
config switch
	option name 'switch0'
	option reset '1'
	option enable_vlan '1'

config switch_vlan
	option device 'switch0'
	option vlan '1'
	option vid '1'
	option ports '0 0 1 2 3 4'

config switch_vlan
	option device 'switch0'
	option vlan '2'
	option vid '2'
	option ports '0t 0t 5'
```

</details>

Then, I applied my custom config:

<details>

```
config switch
	option name 'switch0'
	option reset '1'
	option enable_vlan '1'

config switch_vlan
	option device 'switch0'
	option vlan '1'
	option vid '1'
	option ports '0 0'

config switch_vlan
	option device 'switch0'
	option vlan '2'
	option vid '2'
	option ports '0t 0t 5'

config switch_vlan
	option device 'switch0'
	option vlan '3'
	option ports '0t 0t 1 2'
	option vid '10'

config switch_vlan
	option device 'switch0'
	option vlan '4'
	option ports '0t 0t 3 4'
	option vid '11'
```

</details>

The ordering of `eth1` before `eth0` is important so the VLAN devices are created on `eth0` since the driver filters them out from `eth`.

And, everything seems to work fine! I tested the VLANs directly and by using `tcpdump -e vlan -ni eth0` (and on the other devices and VLAN devices). The only problems are the duplicate CPU ports (they don't seem to cause real problems since they just update to match each other, so it's just a cosmetic issue) and the `eth0.2` device, which doesn't do anything (the driver filters those out so they only appear in `eth1`).

This is my first time doing anything with OpenWRT internals and with this board, so I'm not sure if there are other implications I don't know about.

@robimarko, thoughts?

P.S. Should I move these comments to #3037?

**Comment by mikegleasonjr** [comment]
How can I contribute to buy a coffee to anyone working on the Mikrotik APs, especially the cAPac?

**Comment by kmarzic** [comment]
> I've tried compiling the firmware and I get errors. Have followed the instructions to the letter so I have NFI why
> Can you compile the latest release and upload it?

Hi,
Build is here, if you still need compiled version:
https://github.com/kmarzic/openwrt_mikrotik_cap_ac/releases/tag/mikrotik_cap_ac
I was using:
https://github.com/rusitschka/openwrt-mikrotik-hap-ac2

**Comment by vibornoff** [comment]
> EDIT: heck, if anyone is willing to go to the end of making it work (or at least try hard to) and don't have a device, maybe I am willing to participate in providing a unit for someone... let me know

Mine works too:
- hw switching is fine, vlans configurable with LuCi, no more weird `eth1` device
- all leds may be toggled on and off (aka night mode) with hw button (some minor issue with tx/rx events  for ethernet leds)

Todo:
- fix tx/rx events for ethernet leds
- add POE passthrough; for now it can be turned on or off from command line, no automatic detection of pugged device performed

https://github.com/vibornoff/openwrt/tree/mikrotik-cap-ac-wip

**Comment by mikegleasonjr** [comment]
Thank you very much for you efforts on this. I installed @kmarzic 's build this weekend on my 2 cap ac in my house. I am impressed with wifi 3 roaming so far. Very stable, no problem so far. I have 2 VLANs, one for my private network, another for my guest network.

Again, running an `iperf3` continuously (`-t 1000`) on my phone to a wired server and moving around the house is impressive. It roams between 2.4 and 5ghz between the same access point and it roams easily between access points.. The fact that I just have to select my country and select `auto` as channel selection make it plug and play. Before with RouterOS, I had to play and decrease the 2.4ghz DBI for my `ac` devices to connect to the 5ghz radio. Otherwise it would always connect to the 2.4 ghz radio first.

Again, thank you very much, hope it will make it into the official builds.

**Comment by mikegleasonjr** [comment]
Hi all, I'm having a working set (2) of Cap ACs at home for 2 weeks now on @kmarzic's OpenWrt build.

Rock solid. Roaming works beautifully. Thanks again.

Will we see this integrated into the official builds?

**Comment by Groofster** [comment]
I have been running @kmarzic build on three cap ac's for a while now and they are really rock solid. Both 2.4ghz and 5ghz speeds far exceed that of the routeros (non mu-mimo) firmware. Really happy with fast roaming, when you stay away from wpa3 /owe it really is a very simple setup and my devices roam really, really fast now. Thanks

I do have a question though. I am running all three as dumb Access Points which seems to be working quite nicely but since the cap ac only has PoE in on eth0 and this build designates eth0 as the WAN port, configuring them as simple AP's was not straight forward. I have been locked out of the devices a few times during configuration. Right now I have followed [this openwrt tutorial](https://openwrt.org/docs/guide-user/network/wifi/dumbap) but am not sure what to make of step 6. in that tutorial since there is already a bridge but I can only add eth0 and eth0.1 there.

As stated, everything seems to be working nicely but am not sure if I missed anything. If you need more info or exports of my config, please let me know. More than happy to help to get this build ready to be part of the official openwrt builds. Mikrotik hardware has always had nice hardware for the price, so I think it will be in the interest of openwrt as well to have such a budget option to run openwrt on.

**Comment by mikegleasonjr** [comment]
@alaraun Hi mate, thanks again for the PR. I am running this build at home flawlessly for 11 days uptime now.

@adschm removed the "needs changes" label, so I guess it is ready to be merged as soon as the conflicts are resolved?

**Comment by rusitschka** [comment]
[My docker based 21.02 build](https://github.com/rusitschka/openwrt-mikrotik-hap-ac2) now patches in cAP ac support using [vibornoff's branch](https://github.com/vibornoff/openwrt/tree/mikrotik-cap-ac-wip). Switch works fine without any issues.

However I couldn't figure out how to enable [PoE passthrough](https://github.com/vibornoff/openwrt/commit/88f9deb8524d16b31fe5b40a5d3c017b6684a1b0#r52922214). Any hints?

**Comment by mikegleasonjr** [comment]
> PoE passthrough is working, too:

So you have to put that in the Local Startup script right to be enabled again on the next boot?

**Comment by rusitschka** [comment]
> > PoE passthrough is working, too:
> 
> So you have to put that in the Local Startup script right to be enabled again on the next boot?

Yeah that's what I did.

**Comment by john-tho** [comment]
> > > PoE passthrough is working, too:
> > 
> > 
> > So you have to put that in the Local Startup script right to be enabled again on the next boot?
> 
> Yeah that's what I did.

There is an Openwrt daemon / script that can do this for you, though it still only uses the sysfs interface:
https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/init.d/gpio_switch

Example that adds default disabled config (to `/etc/config/system`) for 760igs:
https://github.com/openwrt/openwrt/blob/b0483b19f981bbe024cd447cfe469765f49cc587/target/linux/ramips/mt7621/base-files/etc/board.d/03_gpio_switches#L10

Maybe:
```
uci set "system.poe_passthrough"="gpio_switch"
uci set "system.poe_passthrough.name"="PoE Passthrough"
uci set "system.poe_passthrough.gpio_pin"="414"
uci set "system.poe_passthrough.value"="1"

uci commit system

/etc/init.d/gpio_switch reload
```

**Comment by robertfoss** [comment]
@adschm What is left to do before merging this?

**Comment by adschm** [comment]
> @adschm What is left to do before merging this?

Find people who review it.

**Comment by zero13th** [comment]
I have cAP ac with 256mb memory, will the memory be recognized automatically?

>  system resource print

                   uptime: 20h39m31s
                  version: 6.48.5 (long-term)
               build-time: Sep/21/2021 13:50:51
         factory-software: 6.40.3
              free-memory: 197.6MiB
             total-memory: 256.0MiB
                      cpu: ARMv7
                cpu-count: 4
            cpu-frequency: 716MHz
                 cpu-load: 1%
           free-hdd-space: 2264.0KiB
          total-hdd-space: 15.3MiB
          architecture-name: arm
               board-name: cAP ac
                 platform: MikroTik


 

> system routerboard print

       routerboard: yes
       board-name: cAP ac
       model: RouterBOARD cAP Gi-5acD2nD
       serial-number: 817xxx
       firmware-type: ipq4000L
       factory-firmware: 3.43
       current-firmware: 6.48.5
       upgrade-firmware: 6.48.5

**Comment by mikegleasonjr** [comment]
@g6094199 I use the builds from @kmarzic from here: https://github.com/kmarzic/openwrt_mikrotik_cap_ac/releases

I know it was mentioned earlier, just wanted to tell you they're rock solid. VLANs (Guest + Main) + PoE Pass-through on the second port is working.

I also installed the irqbalance package, getting 400-500 mbit/s on my Macbook to a wired computer using iperf3:

```
opkg update
opkg install irqbalance
uci set network.globals.packet_steering=1; uci commit network
uci set irqbalance.irqbalance.enabled=1; uci commit
```

**Comment by g6094199** [comment]
@alaraun  @vibornoff 

i've found the "NEW" RBCAPGI-5ACD2ND-XL - cAP XL ac model which seem to have the same HW? just new enclosure with better antennae, imho.

FCC pics:
[Internal-Photos-4773889.pdf](https://github.com/openwrt/openwrt/files/7347143/Internal-Photos-4773889.pdf)

HW:

Product code | RBcAPGi-5acD2nD-XL
-- | --
Architecture | ARM 32bit
CPU | IPQ-4018
CPU core count | 4
CPU nominal frequency | 716 MHz
RouterOS license | 4
Operating System | RouterOS
Size of RAM | 128 MB
Storage size | 16 MB

**Comment by alaraun** [comment]
@adschm please review the PR.

**Comment by f00b4r0** [comment]
This PR will have to be modified when #4679 is merged. It's preferable that #4679 is merged first.

**Comment by alaraun** [comment]
I'm more than happy to commit changes after #4679 gets merged

**Comment by f00b4r0** [comment]
> I'm more than happy to commit changes after #4679 gets merged

I’ll rephrase: merging this PR before #4679 will conflict with #4679 (which removes shipped BDF for Mikrotik devices). Furthermore #4679 is necessary as there is evidence now that devices of a given model need to use the BDF which embedded in the device itself (and not a shipped one) as it can differ from device to device, without any means to programmatically differentiate them.

In other words your PR may work for *your* cAP-AC but not for the next one. Therefore your PR cannot be merged as is. I hope this is clearer. I also hope this makes it clear why we really need #4679 to move forward.

**Comment by hauke** [comment]
Thank you for the patch, I applied it to master in 70eedac9b30a06d0d1445bfa31743440ba535e51.

**Comment by f00b4r0** [comment]
@hauke you do realize this PR will only work on a subset of these devices?

**Pull request closed without merging**
